### PR TITLE
Fix crash when decrypting tokens.

### PR DESCRIPTION
**Pull request by ElektrojungeMS**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* ~[ ] Did you add unit tests?~
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR adds more checks for `nil` and empty strings in case decryption of tokens fail as well as the fixes by @annakocheshkova in 3c74045185a464696f482f12ab855c4bd47c2674.

### CODE REVIEW

**Comment by ElektrojungeMS** [comment]
Some things to think about:

* This PR is against `develop` and will be part of the March release.
* Not sure if I went a bit too far with all the checks and logging.

Any input is welcome!

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m [hunk 1]
```diff
@@ -50,18 +50,20 @@ - (void)createCryptorContexts:(NSString *)keyTag {
...
       CCCryptorCreate(kCCDecrypt, kMSEncryptionAlgorithm, kCCOptionPKCS7Padding, [self.key bytes], kMSCipherKeySize, NULL, &decObj);
-  self.decryptorObject = decObj;
 
   // Log if it failed to create cryptorgraphics contexts.
   if (encStatus != kCCSuccess || decStatus != kCCSuccess) {
     MSLogError([MSAppCenter logTag], @"Could not create cryptor object.");
+    self.cryptorCreated = NO;
+    return;
   }
+  self.encryptorObject = encObj;
```
Comment (clpolet) [inline AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: Yep makes sense to only assign when fully operational :).

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m [hunk 2]
```diff
@@ -100,17 +102,29 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
   }
   NSString *result = nil;
   NSData *dataToDecrypt = [[NSData alloc] initWithBase64EncodedString:string options:0];
```
Comment (clpolet) [inline AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: Have you tried passing a nil string to `initWithBase64EncodedString:` I just hope it doesn't throw. The doc is not clear.

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m [hunk 3]
```diff
@@ -100,17 +102,29 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
   }
   NSString *result = nil;
```
Comment (clpolet) [inline AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: We probably want unit tests to decrypt nil or empty strings.

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m [hunk 4]
```diff
@@ -100,17 +102,29 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
...
-  size_t cipherBufferSize = CCCryptorGetOutputLength(self.decryptorObject, dataToDecrypt.length, true);
-  uint8_t *cipherBuffer = malloc(cipherBufferSize * sizeof(uint8_t));
-  size_t numBytesDecrypted = 0;
-  CCCryptorStatus status = CCCrypt(kCCDecrypt, kMSEncryptionAlgorithm, kCCOptionPKCS7Padding, [self.key bytes], kMSCipherKeySize, nil,
-                                   [dataToDecrypt bytes], dataToDecrypt.length, cipherBuffer, cipherBufferSize, &numBytesDecrypted);
-  if (status != kCCSuccess) {
-    MSLogError([MSAppCenter logTag], @"Error performing decryption.");
+  if (dataToDecrypt) {
+    size_t cipherBufferSize = CCCryptorGetOutputLength(self.decryptorObject, dataToDecrypt.length, true);
+    uint8_t *cipherBuffer = malloc(cipherBufferSize * sizeof(uint8_t));
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: `sizeof(uint8_t)` is **always** `1`. It's a fixed width integer type.

Comment (MatkovIvan) [reply AppCenter/AppCenter/Internals/Util/MSEncrypter.m]:
```suggestion
    uint8_t *cipherBuffer = malloc(cipherBufferSize);
```

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m [hunk 5]
```diff
@@ -277,9 +277,15 @@ - (void)deleteLogsWithBatchId:(NSString *)batchId groupId:(NSString *)groupId {
...
+    if (![encryptedToken isKindOfClass:[NSNull class]] && (encryptedToken.length > 0)) {
       NSString *targetToken = [self.targetTokenEncrypter decryptString:encryptedToken];
-      [log addTransmissionTargetToken:targetToken];
+      if (targetToken) {
+        [log addTransmissionTargetToken:targetToken];
+      } else {
+        MSLogWarning([MSAppCenter logTag], @"Failed to decrypt the target token.");
+      }
+    } else {
+      MSLogWarning([MSAppCenter logTag], @"Encrypted token was NSNull or an empty string.");
```
[the remark this answers is not in the document]
Comment (ElektrojungeMS) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: I'll downgrade it to verbose and adjust the log message

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m (hunk 2 above)
Comment (ElektrojungeMS) [reply AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: Yes. It's actually marked `nonnull` and will throw. That's why `MSEncrypter`'s interface marks APIs as `nonnull`, too. Hence the check in `MSDBLogStorage` at https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1334/files/a6dcf53a682bd747c7f16a46b0817aafd85473ef#diff-8faf03edd2017a6e6cc1e7898cf0ea74R280 for an empty or empty string.

File: AppCenter/AppCenter/Internals/Util/MSEncrypter.m (hunk 3 above)
Comment (ElektrojungeMS) [reply AppCenter/AppCenter/Internals/Util/MSEncrypter.m]: I added a test to check for empty string. Nil string is not supposed to happen (see comment about nullability above), so no need to write a test for it.

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m [hunk 6]
```diff
@@ -277,9 +277,15 @@ - (void)deleteLogsWithBatchId:(NSString *)batchId groupId:(NSString *)groupId {
...
+    if (![encryptedToken isKindOfClass:[NSNull class]] && (encryptedToken.length > 0)) {
       NSString *targetToken = [self.targetTokenEncrypter decryptString:encryptedToken];
-      [log addTransmissionTargetToken:targetToken];
+      if (targetToken) {
+        [log addTransmissionTargetToken:targetToken];
+      } else {
+        MSLogWarning([MSAppCenter logTag], @"Failed to decrypt the target token.");
+      }
+    } else {
+      MSLogVerbose([MSAppCenter logTag], @"Encrypted token was NSNull or an empty string (this is not a 1DS log).");
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: `1DS` -> `CS`.

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m [hunk 7]
```diff
@@ -277,9 +277,15 @@ - (void)deleteLogsWithBatchId:(NSString *)batchId groupId:(NSString *)groupId {
...
     // Deserialize target token.
     NSString *encryptedToken = row[self.targetTokenColumnIndex];
-    if (![encryptedToken isKindOfClass:[NSNull class]]) {
+    if (![encryptedToken isKindOfClass:[NSNull class]] && (encryptedToken.length > 0)) {
       NSString *targetToken = [self.targetTokenEncrypter decryptString:encryptedToken];
-      [log addTransmissionTargetToken:targetToken];
+      if (targetToken) {
+        [log addTransmissionTargetToken:targetToken];
+      } else {
+        MSLogWarning([MSAppCenter logTag], @"Failed to decrypt the target token.");
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: Just warning message? I think we should drop the log with `error` message.

Comment (clpolet) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: I agree.

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m (hunk 6 above)
Comment (clpolet) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: We can just remove the parenthesis I think. It's hypothesis.

Comment (jaelim-ms) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: That works too.

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: I revised it a bit to make it an error when the string is empty only, that shouldn't happen but for is an error for sure.

File: AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m (hunk 6 above)
Comment (clpolet) [reply AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m]: I revised the `if` logic a bit to only print an error when it's an actual error (empty string).

**Pull request merged**